internal/pkg/plan/plan.go

```diff
@@ -0,0 +1,103 @@
+package plan
+
+import (
+	"github.com/merico-dev/stream/internal/pkg/config"
+	"github.com/merico-dev/stream/internal/pkg/plugin"
+	"github.com/merico-dev/stream/internal/pkg/statemanager"
+)
+
+// Plan is an "Actions" plan, it includes all changes should be take with plugins.
+type Plan struct {
+	Changes []*Change
+}
+
+// ExecuteResult holds the execution result with a single plugin.
+// The map's key is a plugin's name.
+// The Result is Succeeded when map's value error is nil.
+type ExecuteResult map[string]error
+
+// Execute will execute all changes included in the Plan and put the execution result into a ExecuteResult channel.
+func (p *Plan) Execute(execResultChan chan<- ExecuteResult) {
+	for _, c := range p.Changes {
+		// We will consider how to execute Action concurrently later.
+		// It involves dependency management.
+		_, err := c.Action(c.Tool)
+		execResultChan <- ExecuteResult{c.Tool.Name: err}
+	}
+	close(execResultChan)
+}
+
+// ActionFunc is a function that Do Action with a plugin. like:
+// plugin.Install() / plugin.Reinstall() / plugin.Uninstall()
+type ActionFunc func(tool *config.Tool) (bool, error)
+
+// Change is a wrapper with a single Tool and its Action should be execute.
+type Change struct {
+	Tool       *config.Tool
+	ActionName statemanager.ComponentAction
+	Action     ActionFunc
+}
+
+// MakePlan takes "State Manager" & "Config" then do some calculate and return a Plan.
+// All actions should be execute is included in this Plan.changes.
+func MakePlan(smgr statemanager.Manager, cfg *config.Config) *Plan {
```

Comment:
This function is too long and not easily readable.

It's over 50 lines with the comments and doesn't fit on the same screen thus needs scrolling up and down.

Needs to be refactored.

Reply:
I think we might not need `statemanager.StatusUninstalled`. If it's uninstalled, we can delete it from the state.

Reply:
1. After figuring out the logic, this piece of function can be refactored to something like:

```go
...

for _, tool := range cfg.Tools {
	state := smgr.GetState(tool.Name)
    if state == nil {
		plan.Changes = append(createInstallAction())
		continue
    }

	switch state.Status {
	case statemanager.StatusSucceeded:
		// do nothing
	case statemanager.StatusFailed:
		plan.Changes = append(createReinstallAction())
	case statemanager.StatusUninstalled:
		plan.Changes = append(createInstallAction())
	}
}

for _, state := range states {
	if state.tool not in cfg.Tools {
		plan.Changes = append(createUninstallAction())
	}
}

...
```

What do you think?

2. Even the two loops can be put into two different functions and by giving them proper names, it can be more readable. For example, "generatePlanAccordingToConfig", "removeNoLongerNeededTools".

3. Some functions (like the createXXXAction) can be moved to a separate file, like helper.go or something like that, to make this file shorter and easier to read without much scrolling.
Some files can be moved to

Reply:
Yeah, this function's logic can be improved, I'll have a try to make it more readable and reasonable.